## Ejercicios de pair programming 2 enero: EDA I

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../datos/tiburon4.csv", index_col=0)

In [3]:
df.drop("Unnamed: 0" , axis= 1, inplace = True) #Eliminamos la columna unnamed:0 que arrastramos de otras ocasiones

In [4]:
df.head()

case_number    year        type    country        area  \
0    2018.06.25  2018.0     Boating        USA  California   
1  2018.06.03.a  2018.0  Unprovoked     BRAZIL  Pernambuco   
2  2018.05.26.b  2018.0  Unprovoked        USA     Florida   
3    2018.05.24  2018.0    Provoked  AUSTRALIA  Queensland   
4  2018.05.13.a  2018.0     Invalid    ENGLAND    Cornwall   

                       location        activity                    name age  \
0   Oceanside, San Diego County        Paddling             Julie Wolfe  57   
1         Piedade Beach, Recife        Swimming  Jose Ernesto da Silva   18   
2  Cocoa Beach, Brevard  County         Walking               Cody High  15   
3               Cairns Aquarium  Feeding sharks                    male  32   
4                Off Land's End         Fishing            Max Berryman  21   

             time          species_  \
0           18h00       White shark   
1  Late afternoon       Tiger shark   
2           17h00    Bull shark, 6'   
3         Morning   Grey reef shark   
4           08h15  Invalid incident   

                                                href      siglo  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
3  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
4  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   

                                              injury  mes deceso genero  
0  No injury to occupant, outrigger canoe and pad...  Jun      N      F  
1                                              FATAL  Jun      Y      M  
2                              Lower left leg bitten  May      N      M  
3  Minor bite to hand by captive shark. PROVOKED ...  May      N      M  
4  Injured by teeth of a dead porbeagle shark he ...  May      N      M

1 Haced una exploración inicial del dataframe:

1.1 ¿Cuál es el número de filas y columnas?


In [9]:
df.shape

(1672, 17)

Tiene 1672 filas y 17 columnas

1.2. ¿Cuál es el porcentaje de valores nulos en cada columna? Presenta los resultados en un dataframe, donde las columnas sean el nombre de las variables y el porcentaje de nulos.

In [13]:
df.isnull().sum() * 100 / df.shape[0] #Vemos el porcentaje de nulos mediante isnull

case_number    0.000000
year           0.000000
type           0.000000
country        0.598086
area           2.751196
location       3.050239
activity       1.854067
name           0.538278
age            9.210526
time           9.569378
species_       7.535885
href           0.000000
siglo          0.000000
injury         0.478469
mes            9.210526
deceso         5.921053
genero         0.837321
dtype: float64

In [16]:
df_nulos = pd.DataFrame(df.isnull().sum() * 100 / df.shape[0]).T #Lo convertimos en dataframe y lo transponemos para mejor visualizacion
df_nulos.head()

case_number  year  type   country      area  location  activity      name  \
0          0.0   0.0   0.0  0.598086  2.751196  3.050239  1.854067  0.538278   

        age      time  species_  href  siglo    injury       mes    deceso  \
0  9.210526  9.569378  7.535885   0.0    0.0  0.478469  9.210526  5.921053   

     genero  
0  0.837321

1.3. ¿Qué tipos de datos hay en cada columna?

In [17]:
df.dtypes

case_number     object
year           float64
type            object
country         object
area            object
location        object
activity        object
name            object
age             object
time            object
species_        object
href            object
siglo           object
injury          object
mes             object
deceso          object
genero          object
dtype: object

1.4. Extraed la información general del dataframe.

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1672 entries, 0 to 1671
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   case_number  1672 non-null   object 
 1   year         1672 non-null   float64
 2   type         1672 non-null   object 
 3   country      1662 non-null   object 
 4   area         1626 non-null   object 
 5   location     1621 non-null   object 
 6   activity     1641 non-null   object 
 7   name         1663 non-null   object 
 8   age          1518 non-null   object 
 9   time         1512 non-null   object 
 10  species_     1546 non-null   object 
 11  href         1672 non-null   object 
 12  siglo        1672 non-null   object 
 13  injury       1664 non-null   object 
 14  mes          1518 non-null   object 
 15  deceso       1573 non-null   object 
 16  genero       1658 non-null   object 
dtypes: float64(1), object(16)
memory usage: 235.1+ KB


2. Cread dos dataframes nuevos, uno que incluya solo las variables categóricas y otro que incluya solo las variables numéricas.

In [5]:
df_numericas = df.select_dtypes(exclude = "object") #Sacamos las columnas numéricas

year
0  2018.0
1  2018.0
2  2018.0
3  2018.0
4  2018.0

In [6]:
df_categoricas = df.select_dtypes(include = "object")  #Sacamos las columnas categoricas
df_categoricas.head()


case_number        type    country        area  \
0    2018.06.25     Boating        USA  California   
1  2018.06.03.a  Unprovoked     BRAZIL  Pernambuco   
2  2018.05.26.b  Unprovoked        USA     Florida   
3    2018.05.24    Provoked  AUSTRALIA  Queensland   
4  2018.05.13.a     Invalid    ENGLAND    Cornwall   

                       location        activity                    name age  \
0   Oceanside, San Diego County        Paddling             Julie Wolfe  57   
1         Piedade Beach, Recife        Swimming  Jose Ernesto da Silva   18   
2  Cocoa Beach, Brevard  County         Walking               Cody High  15   
3               Cairns Aquarium  Feeding sharks                    male  32   
4                Off Land's End         Fishing            Max Berryman  21   

             time          species_  \
0           18h00       White shark   
1  Late afternoon       Tiger shark   
2           17h00    Bull shark, 6'   
3         Morning   Grey reef shark   
4           08h15  Invalid incident   

                                                href      siglo  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
3  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
4  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   

                                              injury  mes deceso genero  
0  No injury to occupant, outrigger canoe and pad...  Jun      N      F  
1                                              FATAL  Jun      Y      M  
2                              Lower left leg bitten  May      N      M  
3  Minor bite to hand by captive shark. PROVOKED ...  May      N      M  
4  Injured by teeth of a dead porbeagle shark he ...  May      N      M

2.1 ¿Cuáles son los principales estádisticos de los dataframes?

In [24]:
df_numericas.describe() #Mediante describe podemos ver los principales estadísticos

year
count  1672.000000
mean   1832.688397
std     528.969901
min       0.000000
25%    1964.000000
50%    1997.000000
75%    2008.000000
max    2018.000000

In [26]:
df_categoricas.describe().T

count unique                                                top  \
case_number  1672   1672                                         2018.06.25   
type         1672      6                                         Unprovoked   
country      1662     95                                                USA   
area         1626    236                                            Florida   
location     1621   1310                   New Smyrna Beach, Volusia County   
activity     1641    404                                            Surfing   
name         1663   1536                                               male   
age          1518     94                                                 19   
time         1512    248                                              11h00   
species_     1546    856                                        White shark   
href         1672   1672  http://sharkattackfile.net/spreadsheets/pdf_di...   
siglo        1672      2                                           siglo-XX   
injury       1664   1173                                              FATAL   
mes          1518     12                                                Aug   
deceso       1573      2                                                  N   
genero       1658      2                                                  M   

             freq  
case_number     1  
type         1432  
country       741  
area          346  
location       73  
activity      426  
name           97  
age            70  
time           74  
species_       76  
href            1  
siglo         887  
injury        154  
mes           162  
deceso       1256  
genero       1435

2.2 Extraed los valores únicos su frecuencia. BONUS presenta los resultados en un dataframe. 

Donde variable es cada una de las variables categóricas que tenemos en el dataset y n_unicos es una lista con los nombres de los valores únicos de cada variable.


In [7]:
df_categoricas["case_number"].unique() 

array(['2018.06.25', '2018.06.03.a', '2018.05.26.b', ..., 'ND.0003',
       'ND.0002', 'ND.0001'], dtype=object)

In [10]:
variables = df_categoricas.columns #Sacamos los nombres de las columnas categóricas y las almacenamos en la variable variables

In [11]:
n_uniques= [] #Creamos una lista vacía para almacenar los n_uniques
for col in variables: #Iteramos por las columnas de las columnas de df_categoricas y por cada una apendeamosa la lista n_uniques el valor del unico de cada columna 
    n_uniques.append(list(df_categoricas[col].unique())) # lista de los n_uniques 

In [12]:
dicc = {'variable': variables, 'n_uniques': n_uniques} #Convertimos las dos listas que tenemos en un diccionario en el que la key es variable y n_uniques y los values son los nombres de las columnas y su lista de valores unicos

In [13]:
dicc

{'variable': Index(['case_number', 'type', 'country', 'area', 'location', 'activity',
        'name', 'age', 'time', 'species_', 'href', 'siglo', 'injury', 'mes',
        'deceso', 'genero'],
       dtype='object'),
 'n_uniques': [['2018.06.25',
   '2018.06.03.a',
   '2018.05.26.b',
   '2018.05.24',
   '2018.05.13.a',
   '2018.04.28.b',
   '2018.04.25.a',
   '2018.04.23',
   '2018.04.22',
   '2018.04.03',
   '2018.03.31',
   '2018.03.14',
   '2018.03.09.b',
   '2018.03.09.a',
   '2018.02.23',
   '2018.02.15',
   '2018.01.12',
   '2017.12.31',
   '2017.11.24',
   '2017.11.13',
   '2017.11.04',
   '2017.10.28',
   '2017.10.22',
   '2017.10.09',
   '2017.10.05',
   '2017.09.24.b',
   '2017.09.13',
   '2017.09.10.a',
   '2017.08.29',
   '2017.08.26.b',
   '2017.08.23',
   '2017.08.10',
   '2017.08.03',
   '2017.07.29.c',
   '2017.07.20.b',
   '2017.07.20.a',
   '2017.07.09',
   '2017.07.07',
   '2017.06.11',
   '2017.06.02',
   '2017.05.03',
   '2017.04.17.a',
   '2017.04.14',
   '2017.04.

In [17]:
final = pd.DataFrame.from_dict(dicc) #Convertimos el diccionario en un dataframe

In [18]:
final.head() #tenemos el dataframe con las variables y los valores únicos

variable                                          n_uniques
0  case_number  [2018.06.25, 2018.06.03.a, 2018.05.26.b, 2018....
1         type  [Boating, Unprovoked, Provoked, Invalid, Sea D...
2      country  [USA, BRAZIL, AUSTRALIA, ENGLAND, COSTA RICA, ...
3         area  [California, Pernambuco, Florida, Queensland, ...
4     location  [Oceanside, San Diego County, Piedade Beach, R...

In [25]:
df_categoricas.describe().T

count unique                                                top  \
case_number  1672   1672                                         2018.06.25   
type         1672      6                                         Unprovoked   
country      1662     95                                                USA   
area         1626    236                                            Florida   
location     1621   1310                   New Smyrna Beach, Volusia County   
activity     1641    404                                            Surfing   
name         1663   1536                                               male   
age          1518     94                                                 19   
time         1512    248                                              11h00   
species_     1546    856                                        White shark   
href         1672   1672  http://sharkattackfile.net/spreadsheets/pdf_di...   
siglo        1672      2                                           siglo-XX   
injury       1664   1173                                              FATAL   
mes          1518     12                                                Aug   
deceso       1573      2                                                  N   
genero       1658      2                                                  M   

             freq  
case_number     1  
type         1432  
country       741  
area          346  
location       73  
activity      426  
name           97  
age            70  
time           74  
species_       76  
href            1  
siglo         887  
injury        154  
mes           162  
deceso       1256  
genero       1435

¿Hay alguna columna que no tenga sentido su tipo de datos?
- En la columna year es un float y debería ser int y la columna 'age' debería ser también un int

¿Hay alguna columna que no tenga sentido su nombre?
- href no da información para analizar

¿Hay alguna columna que no tenga sentido su contenido?
- href no es una columna que nos aporte infomación. 
- La columna de name no nos aporta información para analizar que sea relevante para los ataques de tiburones.


¿Hay alguna columna que tenga demasiadas categorías o valores que nos pueda hacer difícil trabajar con ella?
- Los datos de la columna 'time' no están estandarizados y a veces aparece como una hora concreta, otras veces como afternoon...
- La columna de 'injury' a veces describe todo el suceso, otras veces solo la herida...

